// testcafe/lib/test-run/browser-manipulation-queue.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _utils = require('./commands/utils');

var _type = require('./commands/type');

var _type2 = _interopRequireDefault(_type);

var _message = require('../warnings/message');

var _message2 = _interopRequireDefault(_message);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BrowserManipulationQueue = function () {
    function BrowserManipulationQueue(browserConnection, screenshotCapturer, warningLog) {
        (0, _classCallCheck3.default)(this, BrowserManipulationQueue);

        this.commands = [];
        this.browserId = browserConnection.id;
        this.browserProvider = browserConnection.provider;
        this.screenshotCapturer = screenshotCapturer;
        this.warningLog = warningLog;
    }

    BrowserManipulationQueue.prototype._resizeWindow = function () {
        var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(width, height, currentWidth, currentHeight) {
            return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                    switch (_context.prev = _context.next) {
                        case 0:
                            _context.prev = 0;
                            _context.next = 3;
                            return this.browserProvider.resizeWindow(this.browserId, width, height, currentWidth, currentHeight);

                        case 3:
                            return _context.abrupt('return', _context.sent);

                        case 6:
                            _context.prev = 6;
                            _context.t0 = _context['catch'](0);

                            this.warningLog.addWarning(_message2.default.resizeError, _context.t0.message);
                            return _context.abrupt('return', null);

                        case 10:
                        case 'end':
                            return _context.stop();
                    }
                }
            }, _callee, this, [[0, 6]]);
        }));

        function _resizeWindow(_x, _x2, _x3, _x4) {
            return _ref.apply(this, arguments);
        }

        return _resizeWindow;
    }();

    BrowserManipulationQueue.prototype._takeScreenshot = function () {
        var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(capture) {
            return _regenerator2.default.wrap(function _callee2$(_context2) {
                while (1) {
                    switch (_context2.prev = _context2.next) {
                        case 0:
                            if (this.screenshotCapturer.enabled) {
                                _context2.next = 3;
                                break;
                            }

                            this.warningLog.addWarning(_message2.default.screenshotsPathNotSpecified);
                            return _context2.abrupt('return', null);

                        case 3:
                            _context2.prev = 3;
                            _context2.next = 6;
                            return capture();

                        case 6:
                            return _context2.abrupt('return', _context2.sent);

                        case 9:
                            _context2.prev = 9;
                            _context2.t0 = _context2['catch'](3);

                            this.warningLog.addWarning(_message2.default.screenshotError, _context2.t0.message);
                            return _context2.abrupt('return', null);

                        case 13:
                        case 'end':
                            return _context2.stop();
                    }
                }
            }, _callee2, this, [[3, 9]]);
        }));

        function _takeScreenshot(_x5) {
            return _ref2.apply(this, arguments);
        }

        return _takeScreenshot;
    }();

    BrowserManipulationQueue.prototype.executePendingManipulation = function () {
        var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(driverMsg) {
            var _this = this;

            var command;
            return _regenerator2.default.wrap(function _callee3$(_context3) {
                while (1) {
                    switch (_context3.prev = _context3.next) {
                        case 0:
                            command = this.commands.shift();
                            _context3.t0 = command.type;
                            _context3.next = _context3.t0 === _type2.default.takeScreenshot ? 4 : _context3.t0 === _type2.default.takeScreenshotOnFail ? 7 : _context3.t0 === _type2.default.resizeWindow ? 10 : 13;
                            break;

                        case 4:
                            _context3.next = 6;
                            return this._takeScreenshot(function () {
                                return _this.screenshotCapturer.captureAction({
                                    customPath: command.path,
                                    pageWidth: driverMsg.innerWidth,
                                    pageHeight: driverMsg.innerHeight
                                });
                            });

                        case 6:
                            return _context3.abrupt('return', _context3.sent);

                        case 7:
                            _context3.next = 9;
                            return this._takeScreenshot(function () {
                                return _this.screenshotCapturer.captureError({
                                    screenshotRequired: true,
                                    pageWidth: driverMsg.innerWidth,
                                    pageHeight: driverMsg.innerHeight
                                });
                            });

                        case 9:
                            return _context3.abrupt('return', _context3.sent);

                        case 10:
                            _context3.next = 12;
                            return this._resizeWindow(command.width, command.height, driverMsg.innerWidth, driverMsg.innerHeight);

                        case 12:
                            return _context3.abrupt('return', _context3.sent);

                        case 13:
                            return _context3.abrupt('return', null);

                        case 14:
                        case 'end':
                            return _context3.stop();
                    }
                }
            }, _callee3, this);
        }));

        function executePendingManipulation(_x6) {
            return _ref3.apply(this, arguments);
        }

        return executePendingManipulation;
    }();

    BrowserManipulationQueue.prototype.push = function push(command) {
        this.commands.push(command);
    };

    BrowserManipulationQueue.prototype.removeAllNonServiceManipulations = function removeAllNonServiceManipulations() {
        this.commands = this.commands.filter(function (command) {
            return (0, _utils.isServiceCommand)(command);
        });
    };

    return BrowserManipulationQueue;
}();

exports.default = BrowserManipulationQueue;
module.exports = exports['default'];
