// testcafe/lib/errors/test-run/index.js
'use strict';

exports.__esModule = true;
exports.WindowDimensionsOverflowError = exports.SetNativeDialogHandlerCodeWrongTypeError = exports.UncaughtErrorInNativeDialogHandler = exports.NativeDialogNotHandledError = exports.CurrentIframeIsInvisibleError = exports.CurrentIframeNotFoundError = exports.CurrentIframeIsNotLoadedError = exports.ActionIframeIsNotLoadedError = exports.ActionElementNotIframeError = exports.ActionElementIsNotFileInputError = exports.ActionCanNotFindFileToUploadError = exports.ActionIncorrectKeysError = exports.ActionRootContainerNotFoundError = exports.ActionElementNonContentEditableError = exports.ActionElementNotTextAreaError = exports.ActionElementNonEditableError = exports.ActionAdditionalSelectorMatchesWrongNodeTypeError = exports.ActionAdditionalElementIsInvisibleError = exports.ActionAdditionalElementNotFoundError = exports.ActionSelectorMatchesWrongNodeTypeError = exports.ActionElementIsInvisibleError = exports.ActionElementNotFoundError = exports.ActionSelectorError = exports.ActionUnsupportedDeviceTypeError = exports.ActionUnsupportedUrlProtocolError = exports.ActionStringArrayElementError = exports.ActionStringOrStringArrayArgumentError = exports.ActionPositiveIntegerArgumentError = exports.ActionIntegerArgumentError = exports.ActionStringArgumentError = exports.ActionOptionsTypeError = exports.ActionBooleanOptionError = exports.ActionPositiveIntegerOptionError = exports.ActionIntegerOptionError = exports.ExternalAssertionLibraryError = exports.UncaughtErrorInCustomDOMPropertyCode = exports.UncaughtErrorInClientFunctionCode = exports.UncaughtNonErrorObjectInTestCode = exports.UncaughtErrorInTestCode = exports.UncaughtErrorOnPage = exports.PageLoadError = exports.CantObtainInfoForElementSpecifiedBySelectorError = exports.InvalidSelectorResultError = exports.DomNodeClientFunctionResultError = exports.ClientFunctionExecutionInterruptionError = exports.MissingAwaitError = undefined;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _type = require('./type');

var _type2 = _interopRequireDefault(_type);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Base
//--------------------------------------------------------------------
var TestRunErrorBase = function TestRunErrorBase(type) {
    (0, _classCallCheck3.default)(this, TestRunErrorBase);

    this.type = type;
    this.isTestCafeError = true;
    this.callsite = null;
}; // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------


var ActionOptionErrorBase = function (_TestRunErrorBase) {
    (0, _inherits3.default)(ActionOptionErrorBase, _TestRunErrorBase);

    function ActionOptionErrorBase(type, optionName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionOptionErrorBase);

        var _this = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase.call(this, type));

        _this.optionName = optionName;
        _this.actualValue = actualValue;
        return _this;
    }

    return ActionOptionErrorBase;
}(TestRunErrorBase);

var ActionArgumentErrorBase = function (_TestRunErrorBase2) {
    (0, _inherits3.default)(ActionArgumentErrorBase, _TestRunErrorBase2);

    function ActionArgumentErrorBase(type, argumentName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionArgumentErrorBase);

        var _this2 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase2.call(this, type));

        _this2.argumentName = argumentName;
        _this2.actualValue = actualValue;
        return _this2;
    }

    return ActionArgumentErrorBase;
}(TestRunErrorBase);

// Synchronization errors
//--------------------------------------------------------------------


var MissingAwaitError = exports.MissingAwaitError = function (_TestRunErrorBase3) {
    (0, _inherits3.default)(MissingAwaitError, _TestRunErrorBase3);

    function MissingAwaitError(callsite) {
        (0, _classCallCheck3.default)(this, MissingAwaitError);

        var _this3 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase3.call(this, _type2.default.missingAwaitError));

        _this3.callsite = callsite;
        return _this3;
    }

    return MissingAwaitError;
}(TestRunErrorBase);

// Client function errors
//--------------------------------------------------------------------


var ClientFunctionExecutionInterruptionError = exports.ClientFunctionExecutionInterruptionError = function (_TestRunErrorBase4) {
    (0, _inherits3.default)(ClientFunctionExecutionInterruptionError, _TestRunErrorBase4);

    function ClientFunctionExecutionInterruptionError(instantiationCallsiteName) {
        (0, _classCallCheck3.default)(this, ClientFunctionExecutionInterruptionError);

        var _this4 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase4.call(this, _type2.default.clientFunctionExecutionInterruptionError));

        _this4.instantiationCallsiteName = instantiationCallsiteName;
        return _this4;
    }

    return ClientFunctionExecutionInterruptionError;
}(TestRunErrorBase);

var DomNodeClientFunctionResultError = exports.DomNodeClientFunctionResultError = function (_TestRunErrorBase5) {
    (0, _inherits3.default)(DomNodeClientFunctionResultError, _TestRunErrorBase5);

    function DomNodeClientFunctionResultError(instantiationCallsiteName) {
        (0, _classCallCheck3.default)(this, DomNodeClientFunctionResultError);

        var _this5 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase5.call(this, _type2.default.domNodeClientFunctionResultError));

        _this5.instantiationCallsiteName = instantiationCallsiteName;
        return _this5;
    }

    return DomNodeClientFunctionResultError;
}(TestRunErrorBase);

// Selector errors
//--------------------------------------------------------------------


var InvalidSelectorResultError = exports.InvalidSelectorResultError = function (_TestRunErrorBase6) {
    (0, _inherits3.default)(InvalidSelectorResultError, _TestRunErrorBase6);

    function InvalidSelectorResultError() {
        (0, _classCallCheck3.default)(this, InvalidSelectorResultError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase6.call(this, _type2.default.invalidSelectorResultError));
    }

    return InvalidSelectorResultError;
}(TestRunErrorBase);

var CantObtainInfoForElementSpecifiedBySelectorError = exports.CantObtainInfoForElementSpecifiedBySelectorError = function (_TestRunErrorBase7) {
    (0, _inherits3.default)(CantObtainInfoForElementSpecifiedBySelectorError, _TestRunErrorBase7);

    function CantObtainInfoForElementSpecifiedBySelectorError(callsite) {
        (0, _classCallCheck3.default)(this, CantObtainInfoForElementSpecifiedBySelectorError);

        var _this7 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase7.call(this, _type2.default.cantObtainInfoForElementSpecifiedBySelectorError));

        _this7.callsite = callsite;
        return _this7;
    }

    return CantObtainInfoForElementSpecifiedBySelectorError;
}(TestRunErrorBase);

// Page errors
//--------------------------------------------------------------------


var PageLoadError = exports.PageLoadError = function (_TestRunErrorBase8) {
    (0, _inherits3.default)(PageLoadError, _TestRunErrorBase8);

    function PageLoadError(errMsg) {
        (0, _classCallCheck3.default)(this, PageLoadError);

        var _this8 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase8.call(this, _type2.default.pageLoadError));

        _this8.errMsg = errMsg;
        return _this8;
    }

    return PageLoadError;
}(TestRunErrorBase);

// Uncaught errors
//--------------------------------------------------------------------


var UncaughtErrorOnPage = exports.UncaughtErrorOnPage = function (_TestRunErrorBase9) {
    (0, _inherits3.default)(UncaughtErrorOnPage, _TestRunErrorBase9);

    function UncaughtErrorOnPage(errMsg, pageDestUrl) {
        (0, _classCallCheck3.default)(this, UncaughtErrorOnPage);

        var _this9 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase9.call(this, _type2.default.uncaughtErrorOnPage));

        _this9.errMsg = errMsg;
        _this9.pageDestUrl = pageDestUrl;
        return _this9;
    }

    return UncaughtErrorOnPage;
}(TestRunErrorBase);

var UncaughtErrorInTestCode = exports.UncaughtErrorInTestCode = function (_TestRunErrorBase10) {
    (0, _inherits3.default)(UncaughtErrorInTestCode, _TestRunErrorBase10);

    function UncaughtErrorInTestCode(err, callsite) {
        (0, _classCallCheck3.default)(this, UncaughtErrorInTestCode);

        var _this10 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase10.call(this, _type2.default.uncaughtErrorInTestCode));

        _this10.errMsg = String(err);
        _this10.callsite = callsite;
        return _this10;
    }

    return UncaughtErrorInTestCode;
}(TestRunErrorBase);

var UncaughtNonErrorObjectInTestCode = exports.UncaughtNonErrorObjectInTestCode = function (_TestRunErrorBase11) {
    (0, _inherits3.default)(UncaughtNonErrorObjectInTestCode, _TestRunErrorBase11);

    function UncaughtNonErrorObjectInTestCode(obj) {
        (0, _classCallCheck3.default)(this, UncaughtNonErrorObjectInTestCode);

        var _this11 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase11.call(this, _type2.default.uncaughtNonErrorObjectInTestCode));

        _this11.objType = typeof obj === 'undefined' ? 'undefined' : (0, _typeof3.default)(obj);
        _this11.objStr = String(obj);
        return _this11;
    }

    return UncaughtNonErrorObjectInTestCode;
}(TestRunErrorBase);

var UncaughtErrorInClientFunctionCode = exports.UncaughtErrorInClientFunctionCode = function (_TestRunErrorBase12) {
    (0, _inherits3.default)(UncaughtErrorInClientFunctionCode, _TestRunErrorBase12);

    function UncaughtErrorInClientFunctionCode(instantiationCallsiteName, err) {
        (0, _classCallCheck3.default)(this, UncaughtErrorInClientFunctionCode);

        var _this12 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase12.call(this, _type2.default.uncaughtErrorInClientFunctionCode));

        _this12.errMsg = String(err);
        _this12.instantiationCallsiteName = instantiationCallsiteName;
        return _this12;
    }

    return UncaughtErrorInClientFunctionCode;
}(TestRunErrorBase);

var UncaughtErrorInCustomDOMPropertyCode = exports.UncaughtErrorInCustomDOMPropertyCode = function (_TestRunErrorBase13) {
    (0, _inherits3.default)(UncaughtErrorInCustomDOMPropertyCode, _TestRunErrorBase13);

    function UncaughtErrorInCustomDOMPropertyCode(instantiationCallsiteName, err, prop) {
        (0, _classCallCheck3.default)(this, UncaughtErrorInCustomDOMPropertyCode);

        var _this13 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase13.call(this, _type2.default.uncaughtErrorInCustomDOMPropertyCode, err, prop));

        _this13.errMsg = String(err);
        _this13.property = prop;
        _this13.instantiationCallsiteName = instantiationCallsiteName;
        return _this13;
    }

    return UncaughtErrorInCustomDOMPropertyCode;
}(TestRunErrorBase);

// Assertion errors
//--------------------------------------------------------------------


var ExternalAssertionLibraryError = exports.ExternalAssertionLibraryError = function (_TestRunErrorBase14) {
    (0, _inherits3.default)(ExternalAssertionLibraryError, _TestRunErrorBase14);

    function ExternalAssertionLibraryError(err, callsite) {
        (0, _classCallCheck3.default)(this, ExternalAssertionLibraryError);

        var _this14 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase14.call(this, _type2.default.externalAssertionLibraryError));

        _this14.errMsg = String(err);
        _this14.callsite = callsite;
        return _this14;
    }

    return ExternalAssertionLibraryError;
}(TestRunErrorBase);

// Action parameters errors
//--------------------------------------------------------------------
// Options errors


var ActionIntegerOptionError = exports.ActionIntegerOptionError = function (_ActionOptionErrorBas) {
    (0, _inherits3.default)(ActionIntegerOptionError, _ActionOptionErrorBas);

    function ActionIntegerOptionError(optionName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionIntegerOptionError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionOptionErrorBas.call(this, _type2.default.actionIntegerOptionError, optionName, actualValue));
    }

    return ActionIntegerOptionError;
}(ActionOptionErrorBase);

var ActionPositiveIntegerOptionError = exports.ActionPositiveIntegerOptionError = function (_ActionOptionErrorBas2) {
    (0, _inherits3.default)(ActionPositiveIntegerOptionError, _ActionOptionErrorBas2);

    function ActionPositiveIntegerOptionError(optionName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionPositiveIntegerOptionError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionOptionErrorBas2.call(this, _type2.default.actionPositiveIntegerOptionError, optionName, actualValue));
    }

    return ActionPositiveIntegerOptionError;
}(ActionOptionErrorBase);

var ActionBooleanOptionError = exports.ActionBooleanOptionError = function (_ActionOptionErrorBas3) {
    (0, _inherits3.default)(ActionBooleanOptionError, _ActionOptionErrorBas3);

    function ActionBooleanOptionError(optionName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionBooleanOptionError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionOptionErrorBas3.call(this, _type2.default.actionBooleanOptionError, optionName, actualValue));
    }

    return ActionBooleanOptionError;
}(ActionOptionErrorBase);

var ActionOptionsTypeError = exports.ActionOptionsTypeError = function (_TestRunErrorBase15) {
    (0, _inherits3.default)(ActionOptionsTypeError, _TestRunErrorBase15);

    function ActionOptionsTypeError(actualType) {
        (0, _classCallCheck3.default)(this, ActionOptionsTypeError);

        var _this18 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase15.call(this, _type2.default.actionOptionsTypeError));

        _this18.actualType = actualType;
        return _this18;
    }

    return ActionOptionsTypeError;
}(TestRunErrorBase);

// Arguments errors


var ActionStringArgumentError = exports.ActionStringArgumentError = function (_ActionArgumentErrorB) {
    (0, _inherits3.default)(ActionStringArgumentError, _ActionArgumentErrorB);

    function ActionStringArgumentError(argumentName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionStringArgumentError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB.call(this, _type2.default.actionStringArgumentError, argumentName, actualValue));
    }

    return ActionStringArgumentError;
}(ActionArgumentErrorBase);

var ActionIntegerArgumentError = exports.ActionIntegerArgumentError = function (_ActionArgumentErrorB2) {
    (0, _inherits3.default)(ActionIntegerArgumentError, _ActionArgumentErrorB2);

    function ActionIntegerArgumentError(argumentName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionIntegerArgumentError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB2.call(this, _type2.default.actionIntegerArgumentError, argumentName, actualValue));
    }

    return ActionIntegerArgumentError;
}(ActionArgumentErrorBase);

var ActionPositiveIntegerArgumentError = exports.ActionPositiveIntegerArgumentError = function (_ActionArgumentErrorB3) {
    (0, _inherits3.default)(ActionPositiveIntegerArgumentError, _ActionArgumentErrorB3);

    function ActionPositiveIntegerArgumentError(argumentName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionPositiveIntegerArgumentError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB3.call(this, _type2.default.actionPositiveIntegerArgumentError, argumentName, actualValue));
    }

    return ActionPositiveIntegerArgumentError;
}(ActionArgumentErrorBase);

var ActionStringOrStringArrayArgumentError = exports.ActionStringOrStringArrayArgumentError = function (_ActionArgumentErrorB4) {
    (0, _inherits3.default)(ActionStringOrStringArrayArgumentError, _ActionArgumentErrorB4);

    function ActionStringOrStringArrayArgumentError(argumentName, actualValue) {
        (0, _classCallCheck3.default)(this, ActionStringOrStringArrayArgumentError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB4.call(this, _type2.default.actionStringOrStringArrayArgumentError, argumentName, actualValue));
    }

    return ActionStringOrStringArrayArgumentError;
}(ActionArgumentErrorBase);

var ActionStringArrayElementError = exports.ActionStringArrayElementError = function (_ActionArgumentErrorB5) {
    (0, _inherits3.default)(ActionStringArrayElementError, _ActionArgumentErrorB5);

    function ActionStringArrayElementError(argumentName, actualValue, elementIndex) {
        (0, _classCallCheck3.default)(this, ActionStringArrayElementError);

        var _this23 = (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB5.call(this, _type2.default.actionStringArrayElementError, argumentName, actualValue));

        _this23.elementIndex = elementIndex;
        return _this23;
    }

    return ActionStringArrayElementError;
}(ActionArgumentErrorBase);

var ActionUnsupportedUrlProtocolError = exports.ActionUnsupportedUrlProtocolError = function (_TestRunErrorBase16) {
    (0, _inherits3.default)(ActionUnsupportedUrlProtocolError, _TestRunErrorBase16);

    function ActionUnsupportedUrlProtocolError(argumentName, protocol) {
        (0, _classCallCheck3.default)(this, ActionUnsupportedUrlProtocolError);

        var _this24 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase16.call(this, _type2.default.actionUnsupportedUrlProtocolError));

        _this24.argumentName = argumentName;
        _this24.protocol = protocol;
        return _this24;
    }

    return ActionUnsupportedUrlProtocolError;
}(TestRunErrorBase);

var ActionUnsupportedDeviceTypeError = exports.ActionUnsupportedDeviceTypeError = function (_ActionArgumentErrorB6) {
    (0, _inherits3.default)(ActionUnsupportedDeviceTypeError, _ActionArgumentErrorB6);

    function ActionUnsupportedDeviceTypeError(argumentName, argumentValue) {
        (0, _classCallCheck3.default)(this, ActionUnsupportedDeviceTypeError);
        return (0, _possibleConstructorReturn3.default)(this, _ActionArgumentErrorB6.call(this, _type2.default.actionUnsupportedDeviceTypeError, argumentName, argumentValue));
    }

    return ActionUnsupportedDeviceTypeError;
}(ActionArgumentErrorBase);

// Selector errors


var ActionSelectorError = exports.ActionSelectorError = function (_TestRunErrorBase17) {
    (0, _inherits3.default)(ActionSelectorError, _TestRunErrorBase17);

    function ActionSelectorError(selectorName, errMsg) {
        (0, _classCallCheck3.default)(this, ActionSelectorError);

        var _this26 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase17.call(this, _type2.default.actionSelectorError));

        _this26.selectorName = selectorName;
        _this26.errMsg = errMsg;
        return _this26;
    }

    return ActionSelectorError;
}(TestRunErrorBase);

// Action execution errors
//--------------------------------------------------------------------


var ActionElementNotFoundError = exports.ActionElementNotFoundError = function (_TestRunErrorBase18) {
    (0, _inherits3.default)(ActionElementNotFoundError, _TestRunErrorBase18);

    function ActionElementNotFoundError() {
        (0, _classCallCheck3.default)(this, ActionElementNotFoundError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase18.call(this, _type2.default.actionElementNotFoundError));
    }

    return ActionElementNotFoundError;
}(TestRunErrorBase);

var ActionElementIsInvisibleError = exports.ActionElementIsInvisibleError = function (_TestRunErrorBase19) {
    (0, _inherits3.default)(ActionElementIsInvisibleError, _TestRunErrorBase19);

    function ActionElementIsInvisibleError() {
        (0, _classCallCheck3.default)(this, ActionElementIsInvisibleError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase19.call(this, _type2.default.actionElementIsInvisibleError));
    }

    return ActionElementIsInvisibleError;
}(TestRunErrorBase);

var ActionSelectorMatchesWrongNodeTypeError = exports.ActionSelectorMatchesWrongNodeTypeError = function (_TestRunErrorBase20) {
    (0, _inherits3.default)(ActionSelectorMatchesWrongNodeTypeError, _TestRunErrorBase20);

    function ActionSelectorMatchesWrongNodeTypeError(nodeDescription) {
        (0, _classCallCheck3.default)(this, ActionSelectorMatchesWrongNodeTypeError);

        var _this29 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase20.call(this, _type2.default.actionSelectorMatchesWrongNodeTypeError));

        _this29.nodeDescription = nodeDescription;
        return _this29;
    }

    return ActionSelectorMatchesWrongNodeTypeError;
}(TestRunErrorBase);

var ActionAdditionalElementNotFoundError = exports.ActionAdditionalElementNotFoundError = function (_TestRunErrorBase21) {
    (0, _inherits3.default)(ActionAdditionalElementNotFoundError, _TestRunErrorBase21);

    function ActionAdditionalElementNotFoundError(argumentName) {
        (0, _classCallCheck3.default)(this, ActionAdditionalElementNotFoundError);

        var _this30 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase21.call(this, _type2.default.actionAdditionalElementNotFoundError));

        _this30.argumentName = argumentName;
        return _this30;
    }

    return ActionAdditionalElementNotFoundError;
}(TestRunErrorBase);

var ActionAdditionalElementIsInvisibleError = exports.ActionAdditionalElementIsInvisibleError = function (_TestRunErrorBase22) {
    (0, _inherits3.default)(ActionAdditionalElementIsInvisibleError, _TestRunErrorBase22);

    function ActionAdditionalElementIsInvisibleError(argumentName) {
        (0, _classCallCheck3.default)(this, ActionAdditionalElementIsInvisibleError);

        var _this31 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase22.call(this, _type2.default.actionAdditionalElementIsInvisibleError));

        _this31.argumentName = argumentName;
        return _this31;
    }

    return ActionAdditionalElementIsInvisibleError;
}(TestRunErrorBase);

var ActionAdditionalSelectorMatchesWrongNodeTypeError = exports.ActionAdditionalSelectorMatchesWrongNodeTypeError = function (_TestRunErrorBase23) {
    (0, _inherits3.default)(ActionAdditionalSelectorMatchesWrongNodeTypeError, _TestRunErrorBase23);

    function ActionAdditionalSelectorMatchesWrongNodeTypeError(argumentName, nodeDescription) {
        (0, _classCallCheck3.default)(this, ActionAdditionalSelectorMatchesWrongNodeTypeError);

        var _this32 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase23.call(this, _type2.default.actionAdditionalSelectorMatchesWrongNodeTypeError));

        _this32.argumentName = argumentName;
        _this32.nodeDescription = nodeDescription;
        return _this32;
    }

    return ActionAdditionalSelectorMatchesWrongNodeTypeError;
}(TestRunErrorBase);

var ActionElementNonEditableError = exports.ActionElementNonEditableError = function (_TestRunErrorBase24) {
    (0, _inherits3.default)(ActionElementNonEditableError, _TestRunErrorBase24);

    function ActionElementNonEditableError() {
        (0, _classCallCheck3.default)(this, ActionElementNonEditableError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase24.call(this, _type2.default.actionElementNonEditableError));
    }

    return ActionElementNonEditableError;
}(TestRunErrorBase);

var ActionElementNotTextAreaError = exports.ActionElementNotTextAreaError = function (_TestRunErrorBase25) {
    (0, _inherits3.default)(ActionElementNotTextAreaError, _TestRunErrorBase25);

    function ActionElementNotTextAreaError() {
        (0, _classCallCheck3.default)(this, ActionElementNotTextAreaError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase25.call(this, _type2.default.actionElementNotTextAreaError));
    }

    return ActionElementNotTextAreaError;
}(TestRunErrorBase);

var ActionElementNonContentEditableError = exports.ActionElementNonContentEditableError = function (_TestRunErrorBase26) {
    (0, _inherits3.default)(ActionElementNonContentEditableError, _TestRunErrorBase26);

    function ActionElementNonContentEditableError(argumentName) {
        (0, _classCallCheck3.default)(this, ActionElementNonContentEditableError);

        var _this35 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase26.call(this, _type2.default.actionElementNonContentEditableError));

        _this35.argumentName = argumentName;
        return _this35;
    }

    return ActionElementNonContentEditableError;
}(TestRunErrorBase);

var ActionRootContainerNotFoundError = exports.ActionRootContainerNotFoundError = function (_TestRunErrorBase27) {
    (0, _inherits3.default)(ActionRootContainerNotFoundError, _TestRunErrorBase27);

    function ActionRootContainerNotFoundError() {
        (0, _classCallCheck3.default)(this, ActionRootContainerNotFoundError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase27.call(this, _type2.default.actionRootContainerNotFoundError));
    }

    return ActionRootContainerNotFoundError;
}(TestRunErrorBase);

var ActionIncorrectKeysError = exports.ActionIncorrectKeysError = function (_TestRunErrorBase28) {
    (0, _inherits3.default)(ActionIncorrectKeysError, _TestRunErrorBase28);

    function ActionIncorrectKeysError(argumentName) {
        (0, _classCallCheck3.default)(this, ActionIncorrectKeysError);

        var _this37 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase28.call(this, _type2.default.actionIncorrectKeysError));

        _this37.argumentName = argumentName;
        return _this37;
    }

    return ActionIncorrectKeysError;
}(TestRunErrorBase);

var ActionCanNotFindFileToUploadError = exports.ActionCanNotFindFileToUploadError = function (_TestRunErrorBase29) {
    (0, _inherits3.default)(ActionCanNotFindFileToUploadError, _TestRunErrorBase29);

    function ActionCanNotFindFileToUploadError(filePaths) {
        (0, _classCallCheck3.default)(this, ActionCanNotFindFileToUploadError);

        var _this38 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase29.call(this, _type2.default.actionCanNotFindFileToUploadError));

        _this38.filePaths = filePaths;
        return _this38;
    }

    return ActionCanNotFindFileToUploadError;
}(TestRunErrorBase);

var ActionElementIsNotFileInputError = exports.ActionElementIsNotFileInputError = function (_TestRunErrorBase30) {
    (0, _inherits3.default)(ActionElementIsNotFileInputError, _TestRunErrorBase30);

    function ActionElementIsNotFileInputError() {
        (0, _classCallCheck3.default)(this, ActionElementIsNotFileInputError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase30.call(this, _type2.default.actionElementIsNotFileInputError));
    }

    return ActionElementIsNotFileInputError;
}(TestRunErrorBase);

// Iframe errors


var ActionElementNotIframeError = exports.ActionElementNotIframeError = function (_TestRunErrorBase31) {
    (0, _inherits3.default)(ActionElementNotIframeError, _TestRunErrorBase31);

    function ActionElementNotIframeError() {
        (0, _classCallCheck3.default)(this, ActionElementNotIframeError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase31.call(this, _type2.default.actionElementNotIframeError));
    }

    return ActionElementNotIframeError;
}(TestRunErrorBase);

var ActionIframeIsNotLoadedError = exports.ActionIframeIsNotLoadedError = function (_TestRunErrorBase32) {
    (0, _inherits3.default)(ActionIframeIsNotLoadedError, _TestRunErrorBase32);

    function ActionIframeIsNotLoadedError() {
        (0, _classCallCheck3.default)(this, ActionIframeIsNotLoadedError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase32.call(this, _type2.default.actionIframeIsNotLoadedError));
    }

    return ActionIframeIsNotLoadedError;
}(TestRunErrorBase);

var CurrentIframeIsNotLoadedError = exports.CurrentIframeIsNotLoadedError = function (_TestRunErrorBase33) {
    (0, _inherits3.default)(CurrentIframeIsNotLoadedError, _TestRunErrorBase33);

    function CurrentIframeIsNotLoadedError() {
        (0, _classCallCheck3.default)(this, CurrentIframeIsNotLoadedError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase33.call(this, _type2.default.currentIframeIsNotLoadedError));
    }

    return CurrentIframeIsNotLoadedError;
}(TestRunErrorBase);

var CurrentIframeNotFoundError = exports.CurrentIframeNotFoundError = function (_TestRunErrorBase34) {
    (0, _inherits3.default)(CurrentIframeNotFoundError, _TestRunErrorBase34);

    function CurrentIframeNotFoundError() {
        (0, _classCallCheck3.default)(this, CurrentIframeNotFoundError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase34.call(this, _type2.default.currentIframeNotFoundError));
    }

    return CurrentIframeNotFoundError;
}(TestRunErrorBase);

var CurrentIframeIsInvisibleError = exports.CurrentIframeIsInvisibleError = function (_TestRunErrorBase35) {
    (0, _inherits3.default)(CurrentIframeIsInvisibleError, _TestRunErrorBase35);

    function CurrentIframeIsInvisibleError() {
        (0, _classCallCheck3.default)(this, CurrentIframeIsInvisibleError);
        return (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase35.call(this, _type2.default.currentIframeIsInvisibleError));
    }

    return CurrentIframeIsInvisibleError;
}(TestRunErrorBase);

// Native dialog errors


var NativeDialogNotHandledError = exports.NativeDialogNotHandledError = function (_TestRunErrorBase36) {
    (0, _inherits3.default)(NativeDialogNotHandledError, _TestRunErrorBase36);

    function NativeDialogNotHandledError(dialogType, url) {
        (0, _classCallCheck3.default)(this, NativeDialogNotHandledError);

        var _this45 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase36.call(this, _type2.default.nativeDialogNotHandledError));

        _this45.dialogType = dialogType;
        _this45.pageUrl = url;
        return _this45;
    }

    return NativeDialogNotHandledError;
}(TestRunErrorBase);

var UncaughtErrorInNativeDialogHandler = exports.UncaughtErrorInNativeDialogHandler = function (_TestRunErrorBase37) {
    (0, _inherits3.default)(UncaughtErrorInNativeDialogHandler, _TestRunErrorBase37);

    function UncaughtErrorInNativeDialogHandler(dialogType, errMsg, url) {
        (0, _classCallCheck3.default)(this, UncaughtErrorInNativeDialogHandler);

        var _this46 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase37.call(this, _type2.default.uncaughtErrorInNativeDialogHandler));

        _this46.dialogType = dialogType;
        _this46.errMsg = errMsg;
        _this46.pageUrl = url;
        return _this46;
    }

    return UncaughtErrorInNativeDialogHandler;
}(TestRunErrorBase);

var SetNativeDialogHandlerCodeWrongTypeError = exports.SetNativeDialogHandlerCodeWrongTypeError = function (_TestRunErrorBase38) {
    (0, _inherits3.default)(SetNativeDialogHandlerCodeWrongTypeError, _TestRunErrorBase38);

    function SetNativeDialogHandlerCodeWrongTypeError(actualType) {
        (0, _classCallCheck3.default)(this, SetNativeDialogHandlerCodeWrongTypeError);

        var _this47 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase38.call(this, _type2.default.setNativeDialogHandlerCodeWrongTypeError));

        _this47.actualType = actualType;
        return _this47;
    }

    return SetNativeDialogHandlerCodeWrongTypeError;
}(TestRunErrorBase);

var WindowDimensionsOverflowError = exports.WindowDimensionsOverflowError = function (_TestRunErrorBase39) {
    (0, _inherits3.default)(WindowDimensionsOverflowError, _TestRunErrorBase39);

    function WindowDimensionsOverflowError(callsite) {
        (0, _classCallCheck3.default)(this, WindowDimensionsOverflowError);

        var _this48 = (0, _possibleConstructorReturn3.default)(this, _TestRunErrorBase39.call(this, _type2.default.windowDimensionsOverflowError));

        _this48.callsite = callsite;
        return _this48;
    }

    return WindowDimensionsOverflowError;
}(TestRunErrorBase);
