// testcafe/lib/api/globals/index.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _fixture = require('./fixture');

var _fixture2 = _interopRequireDefault(_fixture);

var _test = require('./test');

var _test2 = _interopRequireDefault(_test);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Globals = function () {
    function Globals(filename) {
        (0, _classCallCheck3.default)(this, Globals);

        this.filename = filename;
        this.currentFixture = null;
        this.collectedTests = [];
    }

    Globals.prototype.setup = function setup() {
        var _this = this;

        Object.defineProperty(global, 'fixture', {
            get: function get() {
                return new _fixture2.default(_this);
            },
            configurable: true
        });

        Object.defineProperty(global, 'test', {
            get: function get() {
                return new _test2.default(_this);
            },
            configurable: true
        });
    };

    Globals.prototype.remove = function remove() {
        delete global.fixture;
        delete global.test;
    };

    return Globals;
}();

exports.default = Globals;
module.exports = exports['default'];
