// testcafe/lib/client-functions/selector-builder/index.js
'use strict';

exports.__esModule = true;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _dedent = require('dedent');

var _dedent2 = _interopRequireDefault(_dedent);

var _clientFunctionBuilder = require('../client-function-builder');

var _clientFunctionBuilder2 = _interopRequireDefault(_clientFunctionBuilder);

var _replicator = require('../replicator');

var _runtime = require('../../errors/runtime');

var _builderSymbol = require('../builder-symbol');

var _builderSymbol2 = _interopRequireDefault(_builderSymbol);

var _message = require('../../errors/runtime/message');

var _message2 = _interopRequireDefault(_message);

var _getCallsite = require('../../errors/get-callsite');

var _getCallsite2 = _interopRequireDefault(_getCallsite);

var _typeAssertions = require('../../errors/runtime/type-assertions');

var _deprecate = require('../../warnings/deprecate');

var _deprecate2 = _interopRequireDefault(_deprecate);

var _observation = require('../../test-run/commands/observation');

var _defineLazyProperty = require('../../utils/define-lazy-property');

var _defineLazyProperty2 = _interopRequireDefault(_defineLazyProperty);

var _addApi = require('./add-api');

var _addApi2 = _interopRequireDefault(_addApi);

var _createSnapshotMethods = require('./create-snapshot-methods');

var _createSnapshotMethods2 = _interopRequireDefault(_createSnapshotMethods);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SelectorBuilder = function (_ClientFunctionBuilde) {
    (0, _inherits3.default)(SelectorBuilder, _ClientFunctionBuilde);

    function SelectorBuilder(fn, options, callsiteNames) {
        (0, _classCallCheck3.default)(this, SelectorBuilder);

        var builderFromSelector = fn && fn[_builderSymbol2.default];
        var builderFromPromiseOrSnapshot = fn && fn.selector && fn.selector[_builderSymbol2.default];
        var builder = builderFromSelector || builderFromPromiseOrSnapshot;

        builder = builder instanceof SelectorBuilder ? builder : null;

        if (builder) {
            fn = builder.fn;

            if (options === void 0 || (typeof options === 'undefined' ? 'undefined' : (0, _typeof3.default)(options)) === 'object') options = (0, _lodash.merge)({}, builder.options, options, { sourceSelectorBuilder: builder });
        }

        return (0, _possibleConstructorReturn3.default)(this, _ClientFunctionBuilde.call(this, fn, options, callsiteNames));
    }

    SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty = function _defineNodeSnapshotDerivativeSelectorProperty(obj, propName, fn) {
        (0, _defineLazyProperty2.default)(obj, propName, function () {
            (0, _deprecate2.default)((0, _getCallsite2.default)('get'), {
                what: 'nodeSnapshot.' + propName,
                useInstead: 'hierarchical selectors (e.g. selector.find())'
            });

            var builder = new SelectorBuilder(function (fnArg) {
                /* eslint-disable no-undef */
                var selectorResult = selector();

                if (selectorResult && typeof selectorResult.then === 'function') return selectorResult.then(function (node) {
                    return fn(node, fnArg);
                });

                return fn(selectorResult, fnArg);
                /* eslint-enable no-undef */
            }, {
                dependencies: {
                    selector: obj.selector,
                    fn: fn
                }
            });

            return builder.getFunction();
        });
    };

    SelectorBuilder.prototype._getCompiledFnCode = function _getCompiledFnCode() {
        // OPTIMIZATION: if selector was produced from another selector and
        // it has same dependencies as source selector, then we can
        // avoid recompilation and just re-use already compiled code.
        var hasSameDependenciesAsSourceSelector = this.options.sourceSelectorBuilder && this.options.sourceSelectorBuilder.options.dependencies === this.options.dependencies;

        if (hasSameDependenciesAsSourceSelector) return this.options.sourceSelectorBuilder.compiledFnCode;

        var code = typeof this.fn === 'string' ? '(function(){return document.querySelectorAll(' + (0, _stringify2.default)(this.fn) + ');});' : _ClientFunctionBuilde.prototype._getCompiledFnCode.call(this);

        if (code) {
            return (0, _dedent2.default)('(function(){\n                    var __f$=' + code + ';\n                    return function(){\n                        var args = __dependencies$.boundArgs || arguments;\n                        return window[\'%testCafeSelectorFilter%\'](__f$.apply(this, args), __dependencies$.filterOptions);\n                    };\n                 })();');
        }

        return null;
    };

    SelectorBuilder.prototype._createInvalidFnTypeError = function _createInvalidFnTypeError() {
        return new _runtime.ClientFunctionAPIError(this.callsiteNames.instantiation, this.callsiteNames.instantiation, _message2.default.selectorInitializedWithWrongType, (0, _typeof3.default)(this.fn));
    };

    SelectorBuilder.prototype._executeCommand = function _executeCommand(args, testRun, callsite) {
        var resultPromise = _ClientFunctionBuilde.prototype._executeCommand.call(this, args, testRun, callsite);

        this._addBoundArgsSelectorGetter(resultPromise, args);

        // OPTIMIZATION: use buffer function as selector not to trigger lazy property ahead of time
        (0, _addApi2.default)(resultPromise, function () {
            return resultPromise.selector;
        }, SelectorBuilder, this.options.customDOMProperties);

        return resultPromise;
    };

    SelectorBuilder.prototype.getFunctionDependencies = function getFunctionDependencies() {
        var dependencies = _ClientFunctionBuilde.prototype.getFunctionDependencies.call(this);
        var text = this.options.text;
        var customDOMProperties = this.options.customDOMProperties;

        if (typeof text === 'string') text = new RegExp((0, _lodash.escapeRegExp)(text));

        return (0, _lodash.merge)({}, dependencies, {
            filterOptions: {
                counterMode: this.options.counterMode,
                collectionMode: this.options.collectionMode,
                index: (0, _lodash.isNil)(this.options.index) ? null : this.options.index,
                text: text
            },

            boundArgs: this.options.boundArgs,
            customDOMProperties: customDOMProperties
        });
    };

    SelectorBuilder.prototype._createTestRunCommand = function _createTestRunCommand(encodedArgs, encodedDependencies) {
        return new _observation.ExecuteSelectorCommand({
            instantiationCallsiteName: this.callsiteNames.instantiation,
            fnCode: this.compiledFnCode,
            args: encodedArgs,
            dependencies: encodedDependencies,
            visibilityCheck: !!this.options.visibilityCheck,
            timeout: this.options.timeout
        });
    };

    SelectorBuilder.prototype._validateOptions = function _validateOptions(options) {
        _ClientFunctionBuilde.prototype._validateOptions.call(this, options);

        if (!(0, _lodash.isNil)(options.visibilityCheck)) (0, _typeAssertions.assertBoolean)(this.callsiteNames.instantiation, '"visibilityCheck" option', options.visibilityCheck);

        if (!(0, _lodash.isNil)(options.text)) (0, _typeAssertions.assertStringOrRegExp)(this.callsiteNames.instantiation, '"text" option', options.text);

        if (!(0, _lodash.isNil)(options.timeout)) (0, _typeAssertions.assertNonNegativeNumber)(this.callsiteNames.instantiation, '"timeout" option', options.timeout);

        if (!(0, _lodash.isNil)(options.index)) (0, _typeAssertions.assertNumber)(this.callsiteNames.instantiation, '"index" option', options.index);
    };

    SelectorBuilder.prototype._getReplicatorTransforms = function _getReplicatorTransforms() {
        var transforms = _ClientFunctionBuilde.prototype._getReplicatorTransforms.call(this);

        transforms.push(new _replicator.SelectorNodeTransform());

        return transforms;
    };

    SelectorBuilder.prototype._addBoundArgsSelectorGetter = function _addBoundArgsSelectorGetter(obj, selectorArgs) {
        var _this2 = this;

        (0, _defineLazyProperty2.default)(obj, 'selector', function () {
            var builder = new SelectorBuilder(_this2.getFunction(), { boundArgs: selectorArgs });

            return builder.getFunction();
        });
    };

    SelectorBuilder.prototype._decorateFunction = function _decorateFunction(selectorFn) {
        _ClientFunctionBuilde.prototype._decorateFunction.call(this, selectorFn);

        (0, _addApi2.default)(selectorFn, function () {
            return selectorFn;
        }, SelectorBuilder, this.options.customDOMProperties);
    };

    SelectorBuilder.prototype._processResult = function _processResult(result, selectorArgs) {
        var snapshot = _ClientFunctionBuilde.prototype._processResult.call(this, result, selectorArgs);

        if (snapshot && !this.options.counterMode) {
            this._addBoundArgsSelectorGetter(snapshot, selectorArgs);

            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getParentNode', function (node) {
                return node ? node.parentNode : node;
            });

            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getChildNode', function (node, idx) {
                return node ? node.childNodes[idx] : node;
            });

            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getChildElement', function (node, idx) {
                if (node.children) return node.children[idx];

                // NOTE: IE doesn't have `children` for non-element nodes =/
                var childNodeCount = node.childNodes.length;
                var currentElIdx = 0;

                for (var i = 0; i < childNodeCount; i++) {
                    if (node.childNodes[i].nodeType === 1) {
                        if (currentElIdx === idx) return node.childNodes[i];

                        currentElIdx++;
                    }
                }

                return null;
            });

            (0, _createSnapshotMethods2.default)(snapshot);
        }

        return snapshot;
    };

    return SelectorBuilder;
}(_clientFunctionBuilder2.default);

exports.default = SelectorBuilder;
module.exports = exports['default'];
